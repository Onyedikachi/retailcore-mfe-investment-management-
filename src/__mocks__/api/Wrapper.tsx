import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";

import { AppStore, RootState, store as S } from "../../config/store";
import { setupListeners } from "@reduxjs/toolkit/dist/query";
import type { PreloadedState } from "@reduxjs/toolkit";
import type { RenderOptions } from "@testing-library/react";
import { PropsWithChildren } from "react";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
  route?: string;
}

interface IWrapper {
  children: React.ReactNode;
}

export function renderWithProviders(
  ui: React.ReactElement | null,
  {
    preloadedState = {},
    store = S(preloadedState),
    route = "/",
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  setupListeners(store.dispatch);
  jest.mock("react-router-dom", () => ({
    BrowserRouter: ({ children }) => <div>{children}</div>,
    Routes: ({ children }) => <div>{children}</div>,
    Route: ({ element }) => element,
  }));
  function Wrappers({ children }: PropsWithChildren<IWrapper>): JSX.Element {
    window.history.pushState({}, "Test page", route);
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrappers, ...renderOptions }) };
}
