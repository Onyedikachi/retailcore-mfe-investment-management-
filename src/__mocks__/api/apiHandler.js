// mswHandlers.js
import { rest } from "msw";
import { response } from "./responses.json";

const baseUrl =
  "https://dev2-retailcore-teams-management-api.dev.bepeerless.co/v1";
const utilBaseUrl = `https://utilities-api.dev.bepeerless.co/v1`;
export const handlers = [
  rest.get(`${baseUrl}/investments`, (req, res, ctx) => {
    // Simulate different responses based on the investmentId
    if (ctx.status == 200) {
      return res(ctx.json(response.getInvestmentsResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.get(`${baseUrl}/quick-link/all`, (req, res, ctx) => {
    // Simulate different responses based on the investmentId
    if (ctx.status == 200) {
      return res(ctx.json(response.getInvestmentsResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.post(`${baseUrl}/investment`, (req, res, ctx) => {
    if (ctx.status == 200) {
      return res(ctx.json(response.getInvestmentsResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.get(`${baseUrl}/investments/members/1`, (req, res, ctx) => {
    // const { filter } = req.params;
    // Simulate different responses based on the investmentId
    if (ctx.status == 200) {
      return res(ctx.json(response.getInvestmentsResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.get(`${baseUrl}/investment/B071`, (req, res, ctx) => {
    // const { filter } = req.params;
    // Simulate different responses based on the investmentId
    if (ctx.status == 200) {
      return res(ctx.json(response.getInvestmentResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.get(`${baseUrl}/investments/analytics`, (req, res, ctx) => {
    // const { filter } = req.params;
    // Simulate different responses based on the investmentId
    if (ctx.status == 200) {
      return res(ctx.json(response.investmentsAnalyticsResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.get(
    `https://dev2-retailcore-teams-management-api.dev.bepeerless.co/v1/Investment/template`,
    (req, res, ctx) => {
      const mockXLSXData = `
   Name, Age
   Alice, 28
   Bob, 24
`;

      const blob = new Blob([mockXLSXData], {
        type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
      });
      if (ctx.status == 200) {
        return res(ctx.json({ file: blob }));
      } else {
        return res(
          ctx.status(404),
          ctx.json({
            error: "Data not found",
          })
        );
      }
    }
  ),

  rest.get(`${baseUrl}/requests`, (req, res, ctx) => {
    // Simulate different responses based on the investmentId
    if (ctx.status == 200) {
      return res(ctx.json(response.getRequestsResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.get(`${baseUrl}/requests/edit/1`, (req, res, ctx) => {
    // Simulate different responses based on the investmentId
    if (ctx.status == 200) {
      return res(ctx.json(response.getRequestsResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.get(`${baseUrl}/requests/edit-bulk/1`, (req, res, ctx) => {
    // Simulate different responses based on the investmentId
    if (ctx.status == 200) {
      return res(ctx.json(response.getRequestsResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.get(`${baseUrl}/requests/analytics`, (req, res, ctx) => {
    // const { filter } = req.params;
    // Simulate different responses based on the investmentId
    if (ctx.status == 200) {
      return res(ctx.json(response.requestAnalyticsResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.get(`${baseUrl}/approve/1`, (req, res, ctx) => {
    // const { filter } = req.params;
    // Simulate different responses based on the investmentId
    return res(ctx.json(response.requestAnalyticsResponse));
    if (ctx.status == 200) {
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.get(`${baseUrl}/location/countries`, (req, res, ctx) => {
    // const { filter } = req.params;
    // Simulate different responses based on the investmentId
    if (ctx.status == 200) {
      return res(ctx.json(response.requestAnalyticsResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.get(`${baseUrl}/location/lga`, (req, res, ctx) => {
    // const { filter } = req.params;
    // Simulate different responses based on the investmentId
    if (ctx.status == 200) {
      return res(ctx.json(response.requestAnalyticsResponse));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),
  rest.post(`${baseUrl}/validate-name`, (req, res, ctx) => {
    if (ctx.status == 200) {
      return res(ctx.json({ success: true }));
    } else {
      return res(
        ctx.status(404),
        ctx.json({
          error: "Data not found",
        })
      );
    }
  }),

  // rest.get(
  //   `https://utilities-api.dev.bepeerless.co/v1/quick-link/all/ProductFactory`,
  //   (req, res, ctx) => {
  //     // const { filter } = req.params;
  //     // Simulate different responses based on the investmentId
  //     if (ctx.status == 200) {
  //       return res(ctx.json(response.requestAnalyticsResponse));
  //     } else {
  //       return res(
  //         ctx.status(500),
  //         ctx.json({
  //           status: "error",
  //           error: "Something went wrong",
  //         })
  //       );
  //     }
  //   }
  // ),


  // rest.post(
  //   `https://utilities-api.dev.bepeerless.co/v1/quick-link`,
  //   (req, res, ctx) => {
  //     // const { filter } = req.params;
  //     // Simulate different responses based on the investmentId
  //     if (ctx.status == 200) {
  //       return res(ctx.json(response.requestAnalyticsResponse));
  //     } else {
  //       return res(
  //         ctx.status(500),
  //         ctx.json({
  //           status: "error",
  //           error: "Something went wrong",
  //         })
  //       );
  //     }
  //   }
  //   ),
    rest.get(`https://retailcore-investment-management-api.dev.bepeerless.co/v1/product/product-details`,
    (req, res, ctx) => {

      if (req.id !== null ) {
        return res(ctx.json(response.getProductDetailResponseOne));
      } else {
        return res(
          ctx.status(500),
          ctx.json({
            status: "error",
            error: "Something went wrong",
          })
        );
      }
    })
    // Add more handlers for different endpoints as needed
];
