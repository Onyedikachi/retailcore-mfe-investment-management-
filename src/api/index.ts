export * from "./linkApi";
export * from "./authApi";
export * from "./investmentApi";
export * from "./accountApi";
