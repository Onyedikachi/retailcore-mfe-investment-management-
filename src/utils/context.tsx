import { createContext, useContext } from "react";

interface ContextProps {
  specificCategory?: string;
  setSpecificCategory?: (e: any) => void;
  category?: any;
  setCategory?: (e: any) => void;
  selected?: { id: number; text: string; value?: string } | null;
  setSelected?: (e: any) => void;

  setStatus?: (e: any) => void;
  status?: string;
  dateData?: any;
  setDateData?: (e: any) => void;
  search?: string;
  setSearch?: (e: any) => void;
  type?: string;
  setType?: (e: any) => void;
  initiator?: string;
  setInitiator?: (e: any) => void;
  setDuration?: (e: any) => void;
  duration?: string;
  isRefreshing?: boolean;
  setRefreshing?: (e: boolean) => void;
  role?: string;
  setRole?: (e: string) => void;
  isDetailOpen?: boolean;
  setDetailOpen?: (e: any) => void;
  isIndividualDetailOpen?: boolean;
  setIndividualDetailOpen?: (e: any) => void;
  detail?: any;
  setDetail?: (e: any) => void;
}

interface branchDeactivationContextProps {
  stage: number;
  setStage: (e: number) => void;
}

interface ISummaryContextProps {
  // productsList: Array<any>;
  branch?: any;
  bulkProductData?: any;
  setBulkProductData?: (e: any) => void;
  requestData?: any;
  bulkQueryFetching?: boolean;
  formData?: any;
  uploadData?: any[];
}

export const AppContext = createContext<{
  role: string;
  setRole: (e) => void;
  permissions: string[];
  currencies?: any;
  setCurrencies?: (e: any) => void;
  userId?: string;
  isChecker?: boolean;
  setIsChecker?: (e: any) => void;
  defaultCurrency?: any;
}>({
  role: "",
  setRole: () => {},
  permissions: [],
  currencies: [],
  setCurrencies: () => {},
  userId: "",
  isChecker: true,
  setIsChecker: () => {},
  defaultCurrency: ""
});

export const CommonContext = createContext<ContextProps>({
  specificCategory: null,
  setSpecificCategory: () => {},
  category: null,
  setCategory: () => {},
  selected: null,
  setSelected: () => {},
  setStatus: () => {},
  status: "",
  dateData: null,
  setDateData: () => {},
  search: "",
  setSearch: () => {},
  type: "",
  setType: () => {},
  initiator: "",
  setInitiator: () => {},
  setDuration: () => {},
  duration: "",
  isRefreshing: false,
  setRefreshing: () => {},
  role: "",
  setRole: () => {},
  isDetailOpen: false,
  setDetailOpen: () => {},
  detail: false,
  setDetail: () => {},
});

export const InvestmentContext = CommonContext;

export const IndividualContext = CommonContext;
interface IOverviewContext {
  overviewTabStats: any;
  setOverviewTabStats: (e: any) => void;
  getStats: () => void;
}

export const OverviewContext = createContext<IOverviewContext>({
  overviewTabStats: null,
  setOverviewTabStats: () => {},
  getStats: () => {},
});



export const ProductDeactivationContext =
  createContext<branchDeactivationContextProps>({
    stage: null,
    setStage: () => {},
  });

export const SummaryContextProps = createContext<ISummaryContextProps>({
  branch: null,
  setBulkProductData: () => {},
  requestData: null,
  bulkQueryFetching: false,
  bulkProductData: null,
  formData: null,
  uploadData: [],
});

// export const defaultOverviewContext = {
//   name: "All Investments",
// };

export function useOverviewContext() {
  const overviewState = useContext(OverviewContext);
  if (overviewState === undefined) {
    throw new Error("useOverviewContext must be used with a dashbord context");
  }

  return overviewState;
}
