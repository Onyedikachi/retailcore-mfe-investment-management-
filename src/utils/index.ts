export * from "./capitalizeFirstLetter";
export * from "./capitalizeObjectKeys";
export * from "./context";
export * from "./getPercentage";
export * from "./removeNullFromObj";
export * from "./requestType";
export * from "./ucObjectKeys";
export * from "./handleDropdown";
export * from "./handleColorState";
export * from "./handleUserView";
export * from "./lowerObjectKeys";
