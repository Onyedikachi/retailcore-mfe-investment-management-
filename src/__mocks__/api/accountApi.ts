import { axiosBaseQuery, getToken } from "@Sterling/shared";
import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { REHYDRATE } from "redux-persist";
import { cleanObject } from "@app/utils/cleanObject";

export const accountApi: any = createApi({
  reducerPath: "accountApi",
  baseQuery: fetchBaseQuery({
    baseUrl: "https://retailcore-accounting-api.dev.bepeerless.co/v1/",
    prepareHeaders: (headers) => {
      const token = getToken();
      if (token) {
        headers.set("authorization", `Bearer ${token}`);
      }
      return headers;
    },
  }),
  keepUnusedDataFor: 0,
  extractRehydrationInfo(action, { reducerPath }) {
    if (action.type === REHYDRATE && action.payload) {
      return action.payload[reducerPath];
    }
  },
  tagTypes: ["Get Ledgers"],
  endpoints: (builder) => ({
    getGlClass: builder.query<any, any>({
      query: () => {
        return {
          url: "accounting/glclass",
          method: "get",
        };
      },
    }),
    getLedgers: builder.query<any, any>({
      query: (data) => {
        return {
          url: `accounting/gl/leaf-ledgers/?${new URLSearchParams(
            cleanObject(data)
          )}`,
          method: "get",
        };
      },
    }),
  }),
});

export const { useGetLedgersQuery, useGetGlClassQuery } = accountApi;
