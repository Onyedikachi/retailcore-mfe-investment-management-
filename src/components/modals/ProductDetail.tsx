import React, { useContext, useState } from "react";
import { IoArrowUndo } from "react-icons/io5";
import { Link } from "react-router-dom";
import { FaBan, FaEdit, FaPlayCircle, FaTimes } from "react-icons/fa";
import moment from "moment";
import ModalLayout from "./Layout";
import BottomBarLoader from "../BottomBarLoader";
import { CustomerCategory, Interval, ProductTypes } from "@app/constants";
import { currencyFormatter } from "@app/utils/formatCurrency";
import { useGetProductDetailQuery } from "@app/api";
import { AppContext } from "@app/utils";

interface Props {
  isOpen: boolean;
  setIsOpen: (isOpen: boolean) => void;
  onCancel?: () => void;
  handleClick?: (e: any, detail: any) => void;
  detail: any;
}

export default function ProductDetail({
  isOpen,
  setIsOpen,
  handleClick,
  detail,
}: Props) {
  const data = [
    "3% for principal between 0 and 200,000",
    "4% for principal between 200,000 and 400,000",
    "3% for principal between 400,000 and 600,000",
    "3% for principal between 0 and 200,000",
    "4% for principal between 200,000 and 400,000",
    "3% for principal between 400,000 and 600,000",
    "3% for principal between 0 and 200,000",
    "4% for principal between 200,000 and 400,000",
    "3% for principal between 400,000 and 600,000",
  ];
  const { data: productData, isLoading } = useGetProductDetailQuery({
    id: detail.id,
  });
  console.log("ðŸš€ ~ file: ProductDetail.tsx:41 ~ productData:", productData)
  const [open, setOpen] = useState(false);
  const { permissions } = useContext(AppContext);

  // React.useEffect(()=>{

  // }, [isSuccess])
  return (
    <ModalLayout isOpen={isOpen} setIsOpen={setIsOpen}>
      <div
        data-testid="product-view"
        className="relative  w-[1218px]  rounded-lg bg-white"
      >
        {!isLoading && (
          <div>
            <div className="flex justify-between items-center pb-6 pt-8 px-16 border-b border-[#CCCCCC] w-full">
              <div className="flex gap-x-5 items-center">
                <h1 className="text-[#636363] font-bold text-2xl uppercase">
                  {productData?.data?.productInfo?.productName || "-"}
                </h1>
                <span
                  className={`${
                    productData?.data?.state === 2
                      ? "text-[#15692A] bg-[#D4F7DC]"
                      : "text-[#1E0A3C] bg-[#E5E5EA]"
                  } px-2 py-[1px] rounded font-medium capitalize`}
                >
                  {productData?.data?.state === 2 ? "Active" : "Inactive"}
                </span>
              </div>
              <button
                onClick={() => setIsOpen(false)}
                className="p-2 outline-none bg-transparent absolute right-6 top-6"
                data-testid="close-btn"
              >
                <FaTimes className="text-[#002266] opacity-60 hover:opacity-50 text-xl" />
              </button>
            </div>
            <div className="grid grid-cols-2 gap-x-5 text-left px-10 py-11">
              <div className="flex flex-col justify-between gap-y-3">
                <div className="p-6 flex flex-col gap-y-[35px] max-h-[463px] overflow-y-auto">
                  {/* {detail?.status === "R" && (
                    <div>
                      <span className="font-bold block mb-[15px]">
                        Reason for Deactivation
                      </span>
                      <span className="font-normal block">
                        {detail?.reason}
                      </span>
                    </div>
                  )} */}
                  <div>
                    <span className="font-bold block mb-[15px]">
                      Product Type
                    </span>
                    <span className="font-normal block uppercase">
                      {ProductTypes.find(
                        (i) => i.id == productData?.data?.productType
                      )?.name || "-"}
                    </span>
                  </div>
                  <div>
                    <span className="font-bold block mb-[15px]">
                      Product Code
                    </span>
                    <span className="font-normal block uppercase">
                      {productData?.data?.productCode || "-"}
                    </span>
                  </div>
                  <div>
                    <span className="font-bold block mb-[15px]">
                      Product Slogan
                    </span>
                    <span className="font-normal block">
                      {productData?.data?.productInfo?.slogan || "-"}
                    </span>
                  </div>

                  <div>
                    <span className="font-bold block mb-[15px]">
                      Product Description
                    </span>
                    <span className="font-normal block">
                      {productData?.data?.productInfo?.description || "-"}{" "}
                    </span>
                  </div>

                  <div>
                    <span className="font-bold block mb-[15px]">
                      Product Life Cycle
                    </span>
                    <span className="font-normal block">
                      {moment(productData?.data?.productInfo?.startDate).format(
                        "DD MMM YYYY"
                      )}{" "}
                      -{" "}
                      {productData?.data?.productInfo?.endDate
                        ? moment(
                            productData?.data?.productInfo?.endDate
                          ).format("DD MMM YYYY")
                        : "Unspecified"}
                    </span>
                  </div>
                  <div>
                    <span className="font-bold block mb-[15px]">Currency</span>
                    <span className="font-normal block">
                      {productData?.data?.productInfo?.currency}{" "}
                    </span>
                  </div>
                  <div>
                    <span className="font-bold block mb-[15px]">
                      Customer Category
                    </span>
                    <span className="font-normal block">
                      {
                        CustomerCategory[
                          productData?.data?.customerEligibility
                            ?.customerCategory
                        ]
                      }{" "}
                    </span>
                  </div>
                </div>
                <div className="border border-[#E5E9EB] rounded-lg py-[35px] px-[30px] flex justify-between items-center">
                  <div className="flex gap-x-6 items-center">
                    {permissions?.includes("CREATE_INVESTMENT_PRODUCT") && (
                      <button
                        data-testid="modify"
                        onClick={() => handleClick("modify", productData?.data)}
                        className={`group flex  items-center whitespace-nowrap  py-[1px] text-base text-[#636363] gap-x-3`}
                      >
                        <FaEdit className="text-[#D4A62F]" /> Modify
                      </button>
                    )}
                    {permissions?.includes(
                      "RE_OR_DEACTIVATE_INVESTMENT_PRODUCT"
                    ) && (
                      <>
                        {productData?.data?.state === 1 ? (
                          <button
                            type="button"
                            data-testid="activate-btn"
                            onClick={() => handleClick("activate", productData?.data)}
                            className={`group flex  items-center whitespace-nowrap  py-[1px] text-base text-[#636363] gap-x-3 outline-none`}
                          >
                            <FaPlayCircle className="text-[#2FB755]" /> Activate
                          </button>
                        ) : (
                          <button
                            data-testid="deactivate-btn"
                            onClick={() => handleClick("deactivate", productData?.data)}
                            className={`group flex  items-center whitespace-nowrap  py-[1px] text-base text-[#636363] gap-x-3 outline-none`}
                          >
                            <FaBan className="text-sterling-red-800" />{" "}
                            Deactivate
                          </button>
                        )}
                      </>
                    )}
                  </div>

                  <Link
                    to={`/product-factory/investment/${encodeURIComponent(
                      "term deposit"
                    )}/process-summary/preview/${productData?.data.id}?category=product`}
                  >
                    <button
                      className={`group flex items-center whitespace-nowrap py-[1px] text-base text-[#636363] gap-x-3 underline outline-none`}
                    >
                      View Activity Log
                    </button>
                  </Link>
                </div>
              </div>
              <div className="border border-[#E5E9EB] rounded-lg py-[25px] px-[30px] h-[593px]">
                <div className="pr-[30px] h-full gap-y-[35px] overflow-y-auto flex flex-col">
                  <div className="">
                    <span className="font-bold block mb-[15px]">
                      Applicable Tenor
                    </span>
                    <span className="font-normal block">
                      {`${
                        productData?.data?.pricingConfiguration
                          ?.applicableTenorMin
                      } ${
                        Interval[
                          productData?.data?.pricingConfiguration
                            ?.applicableTenorMinUnit
                        ]
                      }`}{" "}
                      -{" "}
                      {`${
                        productData?.data?.pricingConfiguration
                          ?.applicableTenorMax
                      } ${
                        Interval[
                          productData?.data?.pricingConfiguration
                            ?.applicableTenorMaxUnit
                        ]
                      }`}
                    </span>
                  </div>
                  <div>
                    <span className="font-bold block mb-[15px]">
                      Applicable Principal
                    </span>
                    <span className="font-normal block">
                      {currencyFormatter(
                        productData?.data?.pricingConfiguration
                          ?.applicablePrincipalMin,
                        productData?.data?.productInfo.currency
                      )}{" "}
                      {productData?.data?.pricingConfiguration
                        ?.applicablePrincipalMax
                        ? `- ${currencyFormatter(
                            productData?.data?.pricingConfiguration
                              ?.applicablePrincipalMax,
                            productData?.data?.productInfo?.currency
                          )}`
                        : "and above"}
                    </span>
                  </div>
                  <div>
                    <span className="font-bold block mb-[15px]">
                      Interest Rate
                    </span>
                    <span className="font-normal block">{productData?.data?.slogan}</span>
                    <button
                      className="text-[#636363]  underline"
                      onClick={() => setOpen(true)}
                    >
                      View more
                    </button>
                  </div>
                  {productData?.data?.liquidation
                    ?.part_AllowPartLiquidation && (
                    <div>
                      <span className="font-bold block mb-[15px]">
                        Part Liquidation
                      </span>
                      <span className="font-normal block">
                        Maximum of{" "}
                        {
                          productData?.data?.liquidation
                            ?.part_MaxPartLiquidation
                        }
                        %{" "}
                        {productData?.data?.liquidation
                          ?.part_LiquidationPenalty === "pay" && (
                          <span> with penalty</span>
                        )}
                      </span>
                    </div>
                  )}
                  {productData?.data?.liquidation
                    ?.early_AllowEarlyLiquidation && (
                    <div>
                      <span className="font-bold block mb-[15px]">
                        Early Liquidation
                      </span>
                      <span className="font-normal block">
                        Allowed{" "}
                        {productData?.data?.liquidation
                          ?.early_LiquidationPenalty === "pay" &&
                          "with penalty"}
                      </span>
                    </div>
                  )}

                  <div>
                    <span className="font-bold block mb-[15px]">
                      Principal Deposit Charge & Tax
                    </span>
                    <div className="flex items-center flex-wrap gap-x-1">
                      <span className="font-normal block">Charges :</span>
                      <span className="font-normal block">
                        Maintenance , Auction
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        )}
        {isLoading && (
          <div className="h-[500px] flex items-center justify-center">
            <div className="spinner-border h-11 w-11 border-t border-danger-500 rounded-full animate-spin"></div>
          </div>
        )}
        <ModalLayout isOpen={open} setIsOpen={setOpen}>
          <div className="px-[30px] pt-[64px] pb-[20px] bg-white w-[400px] rounded-lg relative">
            <ul className="max-h-[345px] overflow-y-auto flex flex-col gap-y-5">
              {data.map((item, idx) => (
                <li key={`${item}-${idx}`}>{item}</li>
              ))}
            </ul>
            <button
              onClick={() => setOpen(false)}
              className="p-2 outline-none bg-transparent absolute right-6 top-6"
              data-testid="close-btn"
            >
              <FaTimes className="text-[#002266] opacity-60 hover:opacity-50 text-xl" />
            </button>
          </div>
        </ModalLayout>
      </div>
    </ModalLayout>
  );
}
