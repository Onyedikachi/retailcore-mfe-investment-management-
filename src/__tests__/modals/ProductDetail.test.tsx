import { act, render, screen, waitFor } from "@testing-library/react"
import { renderWithProviders } from "../../__mocks__/api/Wrapper"
import ProductDetail from "../../components/modals/ProductDetail"
import { InvestmentContext, AppContext } from "../../utils/context";

const details = {
  "id": "0192e82c-3784-4dee-a113-6d113d33eb01",
  
}

class ResizeObserver {
  observe() { }
  unobserve() { }
  disconnect() { }
}

describe("ProductDetail", () => {
  beforeEach(() => {
    window.ResizeObserver = ResizeObserver
  })
  it("Should show product data when available", async () => {
    const val = renderWithProviders(
      <InvestmentContext.Provider value={{}}>
        <AppContext.Provider
          value={{
            permissions: [
              "AUTHORIZE_INVESTMENT_PRODUCT_CREATION_OR_MODIFICATION_REQUESTS",
              "RE_OR_DEACTIVATE_INVESTMENT_PRODUCT"
            ],
            role: "superadmin",
            setRole: jest.fn(),
          }}
        >
          <ProductDetail handleClick={jest.fn()} setIsOpen={jest.fn()} isOpen={true} detail={details} />
        </AppContext.Provider>
      </InvestmentContext.Provider>
    )
    expect(await screen.findByText("Free Loan")).toBeInTheDocument();
    expect(await screen.findByText("LOADNF")).toBeInTheDocument();
    expect(await screen.findByText("tHIS IS A FREE LOAN")).toBeInTheDocument();
    expect(await screen.findByText("NGN")).toBeInTheDocument();
  })
  

  it("Show spinner when loading", async () => {
    const val = renderWithProviders(
      <ProductDetail handleClick={jest.fn()} setIsOpen={jest.fn()} isOpen={true} detail={details} />
    )
    expect(screen.getByTestId("loading-spinner")).toBeInTheDocument();
  })
  
  it ("Shows nothng when setOpen is false", () => {
    const val = renderWithProviders(
      <ProductDetail handleClick={jest.fn()} setIsOpen={jest.fn()} isOpen={false} detail={details} />
    )

    expect(val).toMatchSnapshot();
  })

})

describe("Deactivate button", () => {
  it("Should show activate button when permissions includes 'RE_OR_DEACTIVATE_INVESTMENT_PRODUCT' and state == 2", async () => {
    renderWithProviders(
      <InvestmentContext.Provider value={{}}>
        <AppContext.Provider
          value={{
            permissions: [
              "AUTHORIZE_INVESTMENT_PRODUCT_CREATION_OR_MODIFICATION_REQUESTS",
              "RE_OR_DEACTIVATE_INVESTMENT_PRODUCT"
            ],
            role: "superadmin",
            setRole: jest.fn(),
          }}
        >
          <ProductDetail handleClick={jest.fn()} setIsOpen={jest.fn()} isOpen={true} detail={details} />
        </AppContext.Provider>
      </InvestmentContext.Provider>
    )
    expect(await screen.findByText("Deactivate")).toBeInTheDocument();
  })
})

// describe("Activate button", () => {
//   it("Should show activate button when permissions includes 'RE_OR_DEACTIVATE_INVESTMENT_PRODUCT' and state == 2", async () => {
//     renderWithProviders(
//       <InvestmentContext.Provider value={{}}>
//         <AppContext.Provider
//           value={{
//             permissions: [
//               "AUTHORIZE_INVESTMENT_PRODUCT_CREATION_OR_MODIFICATION_REQUESTS",
//               "RE_OR_DEACTIVATE_INVESTMENT_PRODUCT"
//             ],
//             role: "superadmin",
//             setRole: jest.fn(),
//           }}
//         >
//           <ProductDetail handleClick={jest.fn()} setIsOpen={jest.fn()} isOpen={true} detail={{id : "randomidp"}} />
//         </AppContext.Provider>
//       </InvestmentContext.Provider>
//     )
//     expect(await screen.findByText("Deactivate")).toBeInTheDocument();
//   })
// })