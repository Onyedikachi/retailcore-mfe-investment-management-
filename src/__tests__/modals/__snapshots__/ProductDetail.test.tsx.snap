// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail Renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      aria-hidden="true"
    >
      <div
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
      />
    </div>
    <div
      id="headlessui-portal-root"
    >
      <div
        data-headlessui-portal=""
      >
        <button
          aria-hidden="true"
          data-headlessui-focus-guard="true"
          style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
          type="button"
        />
        <div>
          <div
            aria-modal="true"
            class="relative z-[9999]"
            data-headlessui-state="open"
            data-testid="dialog"
            id="headlessui-dialog-1"
            role="dialog"
          >
            <div
              class="fixed inset-0 bg-black bg-opacity-25 ease-out duration-300 opacity-0"
            />
            <div
              class="fixed inset-0 overflow-y-auto"
              data-testid="Layout"
            >
              <div
                class="flex min-h-full items-center justify-center p-4 text-center"
              >
                <div
                  class="relative transform overflow-hidden rounded-lg text-left align-middle transition-all ease-out duration-300 opacity-0 scale-95"
                >
                  <div
                    class="relative  w-[1218px]  rounded-lg bg-white"
                    data-testid="product-view"
                  >
                    <div
                      class="h-[500px] flex items-center justify-center"
                      data-testid="loading-spinner"
                    >
                      <div
                        class="spinner-border h-11 w-11 border-t border-danger-500 rounded-full animate-spin"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <button
          aria-hidden="true"
          data-headlessui-focus-guard="true"
          style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
          type="button"
        />
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  >
    <div
      style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
